Text tracks are not visible in fullscreen

pvollan opened the pull request:
#### f40d4830b8c1a420d53cc9aabd15c35d4414e711
<pre>
Text tracks are not visible in fullscreen
<a href="https://bugs.webkit.org/show_bug.cgi?id=254473">https://bugs.webkit.org/show_bug.cgi?id=254473</a>
rdar://107230116

Reviewed by Jer Noble.

When the setting &quot;GPU Process: Block Media Layer Re-hosting&quot; is enabled, text tracks are not visible
in fullscreen, since this depends on the ability to create a CA remote layer hosting context in the
WebProcess. This patch works around the requirement to create a CA context in the WebContent process
by sending the caption bitmaps to the UI process, where they are rendered into a CA layer, which is
inserted into the fullscreen CA layer tree. The CA layer tree in the WebContent process is still
being created, since there is some amount of caption code that depends on this in order to get the
caption bounds correct. This patch also sets the flag &quot;GPU Process: Block Media Layer Re-hosting&quot; to
true, which is required to be able to block CARenderServer in the WebContent process&apos; sandbox.
Layout test runs on iOS simulator on the bots showed some flaky timeouts of the test
webrtc/video-rotation.html. Looking at test history, this test has already been flaky on simulator.
This patch is about captions in fullscreen, so I don&apos;t think it will affect this test, although it
cannot be ruled out. I marked the test as flaky. This is being tracked in webkit.org/b/254872.

* LayoutTests/platform/ios-simulator-wk2/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
(WebCore::MediaControlTextTrackContainerElement::updateTextTrackRepresentationIfNeeded):
* Source/WebCore/platform/graphics/TextTrackRepresentation.cpp:
(WebCore::TextTrackRepresentation::create):
* Source/WebCore/platform/graphics/TextTrackRepresentation.h:
* Source/WebCore/platform/graphics/cocoa/TextTrackRepresentationCocoa.h:
* Source/WebCore/platform/graphics/cocoa/TextTrackRepresentationCocoa.mm:
(WebCore::TextTrackRepresentation::create):
(WebCore::TextTrackRepresentationCocoa::representationFactory):
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
(VideoFullscreenInterfaceAVKit::textTrackRepresentationUpdate):
(VideoFullscreenInterfaceAVKit::textTrackRepresentationSetContentsScale):
(VideoFullscreenInterfaceAVKit::textTrackRepresentationSetHidden):
(VideoFullscreenInterfaceAVKit::captionsLayer):
(VideoFullscreenInterfaceAVKit::setCaptionsFrame):
(VideoFullscreenInterfaceAVKit::setupCaptionsLayer):
(VideoFullscreenInterfaceAVKit::removeCaptionsLayer):
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::createLayerHostViewWithID):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationUpdate):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationSetContentsScale):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationSetHidden):
(WebKit::VideoFullscreenManagerProxy::didCleanupFullscreen):
(WebKit::VideoFullscreenManagerProxy::setVideoLayerFrame):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
(WebKit::VideoFullscreenInterfaceContext::rootLayer const):
(WebKit::VideoFullscreenInterfaceContext::setRootLayer):
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenManager::createModelAndInterface):
(WebKit::VideoFullscreenManager::ensureModelAndInterface):
(WebKit::VideoFullscreenManager::setupRemoteLayerHosting):
(WebKit::VideoFullscreenManager::enterVideoFullscreenForVideoElement):
(WebKit::VideoFullscreenManager::requestVideoContentLayer):
(WebKit::VideoFullscreenManager::didCleanupFullscreen):
(WebKit::VideoFullscreenManager::setVideoLayerFrameFenced):
(WebKit::VideoFullscreenManager::updateTextTrackRepresentationForVideoElement):
(WebKit::VideoFullscreenManager::setTextTrackRepresentationContentScaleForVideoElement):
(WebKit::VideoFullscreenManager::setTextTrackRepresentationIsHiddenForVideoElement):

Canonical link: <a href="https://commits.webkit.org/262487@main">https://commits.webkit.org/262487@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d0408a58257d81066f999355ffa2d8cfd36246cd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1648 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1679 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1737 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2570 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1770 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1628 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1745 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1743 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1548 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1663 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1489 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1484 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2406 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1486 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/1470 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1389 "Passed tests") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/1397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1507 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1512 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2567 "run-api-tests-without-change (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1580 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1540 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1372 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/1704 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1458 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1477 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/383 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/411 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1605 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/1746 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/499 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebKit/WebProcess/cocoa/TextTrackRepresentationCocoa.mm:
```diff
@@ -0,0 +1,86 @@
+    auto handle = bitmap->createHandle();
+    if (!handle)
+        return;
+    m_page->send(Messages::VideoFullscreenManagerProxy::TextTrackRepresentationUpdate(*handle));
```
It seems kind of weird to send messages directly to the VideoFullscreenManagerProxy from someplace that's not the VideoFullscreenManager. Instead of passing in `WebPage`, perhaps we could pass in `VideoFullscreenManager` (which WebPage owns) and talk directly to it.

pvollan replied:
> It seems kind of weird to send messages directly to the VideoFullscreenManagerProxy from someplace that's not the VideoFullscreenManager. Instead of passing in `WebPage`, perhaps we could pass in `VideoFullscreenManager` (which WebPage owns) and talk directly to it.

That is a good point. In the latest patch I added the HTMLMediaElement reference parameter to the factory method, which seemed to fit well with keeping the weak WebPage pointer. Are you OK with keeping as is?

jernoble replied:
It still seems very weird to be sending messages directly to the `VideoFullscreenManagerProxy` via the page. You're also not sending a `contextId` in the message, which means there's no way to type the representation to any particular VideoFulllscreenInterface. 

I would say,`VideoFullscreenManager` needs to have the following methods:

```
void updateTextTrackRepresentationForVideoElement(WebCore::HTMLVideoElement&, const ShareableBitmapHandle&);
void setTextTrackRepresentationContentScaleForVideoElement(WebCore::HTMLVideoElement&, float scale);
void setTextTrackRepresentationIsHiddenForVideoElement(WebCore::HTMLVideoElement&, bool hidden);
```

Each of these methods would actually do the sending, after deriving a `contentId` from the passed in element.

Then, the equivalent methods in VideoFullscreenManagerProxy would pass the message contents on to VideoFullscreenInterfaceAVKit, which would actually insert the contents into the layer hierarchy.

jernoble commented on Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
```diff
@@ -127,6 +128,12 @@ static FloatRect inlineVideoFrame(HTMLVideoElement& element)
+        textTrackRepresentation->setPage(page);
+        return textTrackRepresentation;
+    };
+    WebCore::TextTrackRepresentationCocoa::representationFactory() = WTFMove(representationFactory);
```
Shouldn't we be checking the `blockMediaLayerRehostingInWebContentProcess()` setting in here somewhere, so that we have an easy way of turning this path off if there are bugs?

jernoble replied:
Also, I think this is just incorrect code; you're setting a global but passing into the lambda a specific WebPage. It might be okay so long as there is never any more than one WebPage per process, but it has the potential to seriously break if that is ever not true.

I would consider having the factory method take a HTMLMediaElement&. From there you can get the document().page(), convert that to a WebPage, and continue normally.

pvollan replied:
> Shouldn't we be checking the `blockMediaLayerRehostingInWebContentProcess()` setting in here somewhere, so that we have an easy way of turning this path off if there are bugs?

Fixed in latest patch.

pvollan replied:
> Also, I think this is just incorrect code; you're setting a global but passing into the lambda a specific WebPage. It might be okay so long as there is never any more than one WebPage per process, but it has the potential to seriously break if that is ever not true.
> 
> I would consider having the factory method take a HTMLMediaElement&. From there you can get the document().page(), convert that to a WebPage, and continue normally.

That is a good catch! I have addressed this in the latest patch.

jernoble commented on Source/WebKit/WebProcess/cocoa/TextTrackRepresentationCocoa.h:
```diff
@@ -0,0 +1,49 @@
+
+class WebPage;
+
+class TextTrackRepresentationCocoa final : public WebCore::TextTrackRepresentationCocoa {
```
This seems super confusing. Can we just name this class WebTextTrackRepresentationCocoa to disambiguate it from the WebCore class?

pvollan replied:
> This seems super confusing. Can we just name this class WebTextTrackRepresentationCocoa to disambiguate it from the WebCore class?

Done!

jernoble commented on Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
```diff
@@ -335,24 +346,27 @@ static FloatRect inlineVideoFrame(HTMLVideoElement& element)
     interface->setAnimationState(VideoFullscreenInterfaceContext::AnimationType::IntoFullscreen);
 
     bool allowsPictureInPicture = videoElement.webkitSupportsPresentationMode(HTMLVideoElement::VideoPresentationMode::PictureInPicture);
-    
-    if (videoElement.document().settings().blockMediaLayerRehostingInWebContentProcess())
-        m_page->send(Messages::VideoFullscreenManagerProxy::SetupFullscreenWithID(contextId, videoElement.layerHostingContextID(), videoRect, FloatSize(videoElement.videoWidth(), videoElement.videoHeight()), m_page->deviceScaleFactor(), interface->fullscreenMode(), allowsPictureInPicture, standby, videoElement.document().quirks().blocksReturnToFullscreenFromPictureInPictureQuirk()));
-    else {
-        if (!interface->layerHostingContext()->rootLayer()) {
-            auto videoLayer = model->createVideoFullscreenLayer();
-            [videoLayer setDelegate:[WebActionDisablingCALayerDelegate shared]];
-            [videoLayer setName:@"Web Video Fullscreen Layer"];
-            [videoLayer setAnchorPoint:CGPointMake(0, 0)];
-            [videoLayer setPosition:CGPointMake(0, 0)];
-            [videoLayer setBackgroundColor:cachedCGColor(WebCore::Color::transparentBlack).get()];
 
+    if (!interface->rootLayer()) {
```
Why is it ok to create a root layer if we're blocking media layer reposting?

pvollan replied:
> Why is it ok to create a root layer if we're blocking media layer reposting?

I believe it is ok to create a root layer in the WebContent process when blocking blocking media layer re-hosting, since this will not require access to CARenderServer in the sandbox. Ideally, we would also remove the layer tree, but since there is caption code that currently depends on this tree to be present to get the caption bounds correct, I kept it. I think we could address this in a follow-up patch, would you be ok with that?

Thanks for reviewing!

jernoble commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6378,7 +6378,13 @@ void HTMLMediaElement::visibilityStateChanged()
+        return false;
+#if PLATFORM(COCOA)
+    if (document().settings().blockMediaLayerRehostingInWebContentProcess())
+        return true;
```
This seems wrong. We don't want to _always_ paint captions to an image and send to the UI process; only when we're in AVKit fullscreen or PiP mode.

pvollan replied:
Removed, since this was not strictly needed.

jernoble commented on Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
```diff
@@ -598,6 +598,17 @@ - (BOOL)prefersStatusBarHidden
+    interface->captionsLayer().zPosition = FLT_MAX;
+    [interface->captionsLayer() setAnchorPoint:CGPointZero];
+    [interface->captionsLayer() setBounds:CGRectMake(0, 0, initialSize.width(), initialSize.height())];
+    [CATransaction commit];
```
This is fine for now, but I this should all move into VideoFullscreenInterfaceAVKit eventually.

jernoble commented on Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
```diff
@@ -994,6 +1028,8 @@ - (BOOL)prefersStatusBarHidden
 
     m_page->send(Messages::VideoFullscreenManager::SetVideoLayerFrameFenced(contextId, frame, fenceSendRight));
+
+    [interface->captionsLayer() setFrame:CGRectMake(0, 0, frame.width(), frame.height())];
```
...because this should be the responsibility of the LayerHostView to perform when is own size changes.

pvollan committed:
Text tracks are not visible in fullscreen
https://bugs.webkit.org/show_bug.cgi?id=254473
rdar://107230116

Reviewed by Jer Noble.

When the setting "GPU Process: Block Media Layer Re-hosting" is enabled, text tracks are not visible
in fullscreen, since this depends on the ability to create a CA remote layer hosting context in the
WebProcess. This patch works around the requirement to create a CA context in the WebContent process
by sending the caption bitmaps to the UI process, where they are rendered into a CA layer, which is
inserted into the fullscreen CA layer tree. The CA layer tree in the WebContent process is still
being created, since there is some amount of caption code that depends on this in order to get the
caption bounds correct. This patch also sets the flag "GPU Process: Block Media Layer Re-hosting" to
true, which is required to be able to block CARenderServer in the WebContent process' sandbox.
Layout test runs on iOS simulator on the bots showed some flaky timeouts of the test
webrtc/video-rotation.html. Looking at test history, this test has already been flaky on simulator.
This patch is about captions in fullscreen, so I don't think it will affect this test, although it
cannot be ruled out. I marked the test as flaky. This is being tracked in webkit.org/b/254872.

* LayoutTests/platform/ios-simulator-wk2/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
(WebCore::MediaControlTextTrackContainerElement::updateTextTrackRepresentationIfNeeded):
* Source/WebCore/platform/graphics/TextTrackRepresentation.cpp:
(WebCore::TextTrackRepresentation::create):
* Source/WebCore/platform/graphics/TextTrackRepresentation.h:
* Source/WebCore/platform/graphics/cocoa/TextTrackRepresentationCocoa.h:
* Source/WebCore/platform/graphics/cocoa/TextTrackRepresentationCocoa.mm:
(WebCore::TextTrackRepresentation::create):
(WebCore::TextTrackRepresentationCocoa::representationFactory):
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
(VideoFullscreenInterfaceAVKit::textTrackRepresentationUpdate):
(VideoFullscreenInterfaceAVKit::textTrackRepresentationSetContentsScale):
(VideoFullscreenInterfaceAVKit::textTrackRepresentationSetHidden):
(VideoFullscreenInterfaceAVKit::captionsLayer):
(VideoFullscreenInterfaceAVKit::setCaptionsFrame):
(VideoFullscreenInterfaceAVKit::setupCaptionsLayer):
(VideoFullscreenInterfaceAVKit::removeCaptionsLayer):
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::createLayerHostViewWithID):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationUpdate):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationSetContentsScale):
(WebKit::VideoFullscreenManagerProxy::textTrackRepresentationSetHidden):
(WebKit::VideoFullscreenManagerProxy::didCleanupFullscreen):
(WebKit::VideoFullscreenManagerProxy::setVideoLayerFrame):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
(WebKit::VideoFullscreenInterfaceContext::rootLayer const):
(WebKit::VideoFullscreenInterfaceContext::setRootLayer):
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenManager::createModelAndInterface):
(WebKit::VideoFullscreenManager::ensureModelAndInterface):
(WebKit::VideoFullscreenManager::setupRemoteLayerHosting):
(WebKit::VideoFullscreenManager::enterVideoFullscreenForVideoElement):
(WebKit::VideoFullscreenManager::requestVideoContentLayer):
(WebKit::VideoFullscreenManager::didCleanupFullscreen):
(WebKit::VideoFullscreenManager::setVideoLayerFrameFenced):
(WebKit::VideoFullscreenManager::updateTextTrackRepresentationForVideoElement):
(WebKit::VideoFullscreenManager::setTextTrackRepresentationContentScaleForVideoElement):
(WebKit::VideoFullscreenManager::setTextTrackRepresentationIsHiddenForVideoElement):

Canonical link: https://commits.webkit.org/262487@main
